### ISSUE TITLE
Changing batch size and cache

### ISSUE THREAD

**Issue by dvianisoho** [turboderp/exllamav2#131]
I am using batch inference to work through a large number of prompts and everything is working great, however my prompts are all different lengths and I have limited VRAM.  I'd like to be able to run all the < 1024 token prompts in batches of 8, the >1024 but < 2048 in batches of 4, etc.   What is the best way to release the cache and update the batch sizes?    I've tried calling del on the model, generator, and cache objects, and then recreating them, but when I update the config, and create the new cache, and then call model.load_autosplit(cache), I get the following error:

  model.load_autosplit(self.cache)
  File "/tmp/exllamav2/exllamav2/model.py", line 348, in load_autosplit
    hidden_state = module.forward(hidden_state, cache = cache, attn_mask = attn_mask, past_len = past_len, loras = loras)
  File "/tmp/exllamav2/exllamav2/attn.py", line 285, in forward
    k_states = batch_keys.narrow(0, 0, batch_size).narrow(1, past_len, q_len)
RuntimeError: start (0) + length (2048) exceeds dimension size (968).

If I just do model.unload(), recreate the cache, and load the model again, I get a little further, but get the following error when generator.warmup() runs:
 File "/tmp/exllamav2/exllamav2/generator/base.py", line 37, in warmup
    self.model.forward(input_ids, cache = None, input_mask = None, preprocess_only = True)
  File "/tmp/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/tmp/exllamav2/exllamav2/model.py", line 509, in forward
    result, last_state = self._forward(input_ids = input_ids,
  File "/tmp/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/tmp/exllamav2/exllamav2/model.py", line 650, in _forward
    head_padding = self.modules[-1].padding
IndexError: list index out of range

I've also tried clearing the CUDA cache and doing GC after deleting the objects but still get the same errors.   

Followup question - should I expect to see better performance by shortening context length to run more batches?  Is there a better strategy to get higher throughput on many prompts of various lengths? 

Any help is appreciated and thanks for the great work!

**Comment by turboderp** [comment]
You shouldn't need to unload the model at all. The cache is self-contained, so as long as you clear all references to it (and the generator which keeps its own reference), the garbage collector should be able to free it. You can try this to force garbage collection and freeing up the CUDA cache, though it technically shouldn't be needed:

```python
import gc, torch

gc.collect()
torch.cuda.empty_cache()
```

**Comment by dvianisoho** [comment]
Thank you.  Not deleting the model and just killing the generator and cache worked, so problem solved.  That said, if the model IS deleted as well, then this error occurs, which does seem odd.   Try the following:

```
from exllamav2 import ExLlamaV2, ExLlamaV2Config, ExLlamaV2Cache, ExLlamaV2Tokenizer
from exllamav2.generator import ExLlamaV2BaseGenerator
config = ExLlamaV2Config()
config.model_dir = "/var/data/summarizer/models/vic15-13-128-false"
config.prepare()
model = ExLlamaV2(config)
cache = ExLlamaV2Cache(model, lazy=True, batch_size=2, max_seq_len=2048)
model.load_autosplit(cache)
tokenizer=ExLlamaV2Tokenizer(config)        
generator = ExLlamaV2BaseGenerator(model, cache, tokenizer)
generator.warmup()
del generator
del cache
del model
model = ExLlamaV2(config)
cache = ExLlamaV2Cache(model, lazy=True, batch_size=4, max_seq_len=700)
generator = ExLlamaV2BaseGenerator(model, cache, tokenizer)
generator.warmup()
```
Which is now showing a different error from the two above:
```
line 608, in _forward
    device = _torch_device(module.device_idx)
AttributeError: 'ExLlamaV2Embedding' object has no attribute 'device_idx'
```
This happens even when garbage collection and clearing cuda cache is done.

**Comment by turboderp** [comment]
Hmm, you're not calling `load` or `load_autosplit` on the model after creating the new instance. Could that be it?

**Comment by dvianisoho** [comment]
Thank you again and you are of course correct - my mistake (it was late).    Everything is working as expected now.   My actual issue all along was `max_seq_len` in the model config didn't match `max_seq_len` in the cache.   This is what was causing the  original error during `load_autosplit`.   

I also found that `max_input_len` in the config needs to be set to` min(max_seq_len, 2048)` for the model I am using (llama2 GPTQ).   Things work fine if `max_input_len` is set to 2048, and context is longer, but stop working if `max_seq_len` is shorter than `max_input_len`, which felt a little counterintuitive (at least to me but I am sure there is  good reason).  Is this making sense?

Apologies for not responding sooner.   Every time I'd do testing and come back here to update this post with a new issue, it seems you have already uploaded a fix (specifically the memory leak,  batch performance issue, and the multicache example)!

I've been doing some performance testing with various batching / caching strategies with real world data and will share my finding here when they are complete.   My use case is large batches of prompts (10k +) of various lengths (500 - 4000 tokens) and generation lengths of 50 - 250 tokens using a 13GB GPTQ model on a 16GB card.    FYI, of all the things I've tried so far, it's been pretty tough to do a whole lot better than just running prompts through one at a time.   If you have any ideas about how to squeeze performance out of a scenario like that I'm all ears.

**Comment by turboderp** [comment]
Right, the `max_input_len` is not very intuitively named. When you do a forward pass on a long sequence, like a very long prompt, it's split into chunks of a fixed max length, which limits the size of the fixed temporary tensors that need to be allocated as scratch space. Combined with `max_attn_size` it also limits the size of the attention weights matrix to keep the quadratic scaling under control.

I hadn't considered what would happen when the models `max_seq_len` is less than `max_input_len` but I will do some tests.
